WordPress/gutenberg
# [RNMobile] Embed Block Bottom Sheet to set a link

## Description


This PR implements the Embed Block Bottom Sheet to set a link.

Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3436

### Use of `onClose` and `onDismiss` callbacks from the bottom sheet component

It's not very clear the difference between the `onClose` and `onDismiss` callbacks, so for the purpose of this PR, I'd like to briefly describe when these callbacks are triggered and how they're used in the embed sheet.

#### `onClose`
This callback is triggered by the actions listed below, as far as I checked, looks like the callback is intended for requesting the parent to handle the close flow.

One thing I'd like to note is that this being called multiple times in some cases, I'm not sure if this behavior is intended or not, so it would be nice that we investigate it deeper in the future. As an example, when tapping the backdrop element when the modal is visible, the callback is triggered a first time, and then right after, as the modal is dismissed, it's triggered for a second time.

This is caused by the `onDismiss` function from the component that for some reason, calls `onCloseBottomSheet` that triggers the `onClose` callback:
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L310-L318
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L340-L351

**Actions that trigger `onClose` callback:**
- Tapping the backdrop element (related to [`onBackdropPress` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L497)).
- Pressing the back button (only Android) (related to [`onBackButtonPress` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L498)).
- Swiping down the modal (related to [`onSwipeComplete callback`](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L499)).
- Perform the escape gesture (related to [`onAccessibilityEscape` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L513)).
- After the modal is dismissed ([related code](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L310-L318)).

#### `onDismiss`
This is triggered when the modal has been already dismissed/hidden, which only happens once in the closing flow.

#### How are they used in the embed sheet

- `onClose`: As is intended for hiding the modal, we're using it just to update the internal state of the placeholder component and set the visibility value to `false`. As mentioned earlier, this can be triggered multiple times but since we're only updating the internal state this is not an issue.
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-placeholder.native.js#L80
- `onDismiss`: It's being used for handling the URL attribute update. The reason for using this callback instead of `onClose` is because we have the guarantee that will be called only once.
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-bottom-sheet.native.js#L36-L46

## How has this been tested?




1. Add the Embed block
2. Bottom sheet will open and link field will be focused
3. If there is a link in the clipboard it should be auto-pasted
4. Type a link if empty
5. Can click on the “Learn more about embeds” link at the bottom to open the support article in the browser
6. Close the bottom sheet
7. If the link is not a valid URL, a warning message is shown

## Missing features

- ~~Can re-open by pressing on the embed block placeholder. 
Depends on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3273.~~
- Can open by pressing the pen icon in the toolbar when an existing embed block is selected. 
Depends on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3276 
- Edit/replace the current URL placeholder, this will be addressed on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3285

## Screenshots 

iOS Light Mode | iOS Dark Mode 
--------|-------
<img width="300" src="https://user-images.githubusercontent.com/14905380/123625333-f3b88f80-d80f-11eb-988d-ab644ba35f04.png">|<img width="300" src="https://user-images.githubusercontent.com/14905380/123625338-f5825300-d80f-11eb-93fa-0d8fd4ffa27a.png">

Android Light Mode | Android Dark Mode 
--------|-------
<img width="300" src="https://user-images.githubusercontent.com/14905380/123625778-688bc980-d810-11eb-8196-6dac86449137.jpg">|<img width="300" src="https://user-images.githubusercontent.com/14905380/123625787-6aee2380-d810-11eb-9973-442e8a44f77c.jpg">

iOS video capture|Android video capture
--|--
<video src=https://user-images.githubusercontent.com/14905380/123624724-362d9c80-d80f-11eb-9005-6182a843235c.mp4>|<video src=https://user-images.githubusercontent.com/14905380/123626445-2adb7080-d811-11eb-93a8-a8a35835482c.mp4>

## Types of changes


New feature (non-breaking change which adds functionality)


## Checklist:
